### DISCUSSION TITLE
Is it possible to pre-create cluster-admin.crt from custom client-ca?

### DISCUSSION THREAD

**Discussion by mkoperator** [k3s-io/k3s#6638] [Q&A] [answered]
Working with a use-case where I cannot access the k3s master nodes to pull remote cluster-admin access once they are created or create tokens for service accounts. 

I tried generating the client-ca and then using that to create a client-admin, however, my self generated client-admin.crt and key prevent K3s api from becoming available k8s 1.24. I see work is being done in #6615 based on issue #3650 to address the root certificates, but not sure how we get from there to pre-programmed remote access via either an SA or the cluster-admin. Is my approach here correct or is there another way to accomplish this?

**Comment by brandond** [comment, marked as answer]
Follow the instructions at https://github.com/k3s-io/k3s/issues/1868#issuecomment-639690634, then sign your own certificate using the client CA. You can use the same subject as the certificate in the k3s.yaml admin kubeconfig (`CN=system:admin,OU=system:masters`) , or set up your own RBAC once the cluster is up and use a different group if you'd like.

**Reply by mkoperator** [reply]
I think that is what i am doing here: https://github.com/thecrazyrussian/tf-k3s-client-admin/blob/main/main.tf
then copying the `client-ca.crt`, `client-ca.key`, `client-admin.crt`, and `client-admin.key` into `/var/lib/rancher/k3s/tls/` prior to k3s spinning up first time. When I just do the client-ca, the cluster starts up fine and generates new client-admin key that I can use (if I had box access). However, if I also deploy the terraform generated client-admin cert and key K3s service hangs on spin up with "Waiting for API." (k3s 1.24)

**Reply by brandond** [reply]
Don't do that. Only place the CA certs and keys on disk beforehand. If you have SOME of the other files on disk (ie, the client certs but not the kubeconfig, or other variations thereof) it won't reliably create the other files and will fail to start. There's no reason to place the admin client cert and key on the node anyway.

**Reply by mkoperator** [reply]
You are a life savor! @brandond thanks so much, bud! turns out, reason my stuff didnt work is because I confused my private keys.  So now it works perfectly self generating the ca and then key with proper subject. Able to connect to cluster just fine. Sorry about the effort, all me on this one.
